### ISSUE TITLE
Fail to recognize langchain.llms.LlamaCpp llm

### ISSUE THREAD

**Issue by hhy-joseph** [gventuri/pandas-ai#496]
### 🐛 Describe the bug

Want to use langchain.llms.LlamaCpp as llm, but return LLMNotFoundError in PandasAI class.  Am I doing something wrong?

```python
llama = LlamaCpp(
    model_path="./wizardcoder-python-34b-v1.0.Q4_K_S.gguf",
    verbose=False,n_ctx=4096*4,
    n_gpu_layers=20,
    n_batch=20,
    streaming=True, 
)
llama_pandasai = PandasAI(llm=llama)
```

---->  llama_pandasai = PandasAI(llm=llama)

File ~/miniconda3/envs/data_agent/lib/python3.9/site-packages/pandasai/__init__.py:145, in PandasAI.__init__(self, llm, conversational, verbose, enforce_privacy, save_charts, save_charts_path, enable_cache, middlewares, custom_whitelisted_dependencies, enable_logging, non_default_prompts, callback)
    118 """
    119 __init__ method of the Class PandasAI
    120 
   (...)
    138     Used to override specific types of prompts. Defaults to None.
    139 """
    141 # configure the logging
    142 # noinspection PyArgumentList
    143 # https://stackoverflow.com/questions/61226587/pycharm-does-not-recognize-logging-basicconfig-handlers-argument
--> 145 self._config = Config(
    146     conversational=conversational,
    147     verbose=verbose,
    148     enforce_privacy=enforce_privacy,
    149     save_charts=save_charts,
    150     save_charts_path=save_charts_path,
    151     enable_cache=enable_cache,
    152     middlewares=middlewares,
...
     32     if llm is None or not isinstance(llm, LLM or LangchainLLM):
---> 33         raise LLMNotFoundError("LLM is required")
     34     return llm

LLMNotFoundError: LLM is required

**Comment by mspronesti** [comment]
PandasAI only supports a small subset of "trusted" LLM. `LlamaCpp` is not part of them. You can however use `LlamaCpp` with pandasai importing it from langchain

```py
from langchain.llms import LlamaCpp

llama = LlamaCpp(...)
```

For more, I recommend reading the [docs](https://docs.pandas-ai.com/en/latest/LLMs/langchain).

**Comment by hhy-joseph** [comment]
my llama object is from langchain.llms import LlamaCpp right now. So I'm not sure why PandasAI couldn't recognize  this langchain.llms.LlamaCpp.

**Comment by hhy-joseph** [comment]
my llama object is from langchain.llms import LlamaCpp right now. So I'm not sure why PandasAI couldn't recognize this langchain.llms.LlamaCpp.

I have done pip install pandasai[langchain]

**Comment by mspronesti** [comment]
My bad, I missed the fact that you were importing it from `langchain`, sorry for that. Going to the actual question, I think I managed to reproduce the error. There's probably a bug where the conversion is performed. 
I tried to convert the langchain llm to a pandasai llm manually and it did work.

Try this:
```py
from pandasai.llm import LangchainLLM
from pandasai import PandasAI
from langchain.llms import LlamaCpp #Use Langchain llm

llama = LlamaCpp(
    model_path="./wizardcoder-python-34b-v1.0.Q4_K_S.gguf",
    verbose=False,n_ctx=4096*4,
    n_gpu_layers=20,
    n_batch=20,
    streaming=True, 
)

llm = LangchainLLM(llama)
llama_pandasai = PandasAI(llm=llama)
```

**Comment by hhy-joseph** [comment]
```python
llm = LangchainLLM(llama)
llama_pandasai = PandasAI(llm=llm )
```

adjusted the variable and it worked! 

Thank you!

**Comment by emonzies** [comment]
Same error raised (LLMNotFoundError: LLM is required) with AzureOpenAI langchain LLM. It still works with version 0.8, but no longer with pandasai version 1.+

**Comment by gventuri** [comment]
@hhy-joseph @emonzies does the same happen if you use the new `SmartDataframe` or `SmartaDatalake`?

`PandasAI` class is handled only for backward compatibility and will be discontinued soon. Check out the docs and let me know if it fixes: https://docs.pandas-ai.com/en/latest/

**Comment by hhy-joseph** [comment]
This fail:
```python
from pandasai import SmartDataframe
llama_lc = LlamaCpp(
    model_path="./wizardcoder-python-34b-v1.0.Q4_K_S.gguf",
    callback_manager=callback_manager,
    verbose=False,n_ctx=4096*4,
    n_gpu_layers=20,
    n_batch=20,
    streaming=True, 
)
dd = SmartDataframe(df, config={"llm": llama_lc})

dd.chat("scatter plot of sepallength vs sepalwidth color by class")
```

LLMNotFoundError                          Traceback (most recent call last)
Cell In[27], line 11
      2 llama_lc = LlamaCpp(
      3     model_path="./wizardcoder-python-34b-v1.0.Q4_K_S.gguf",
      4     callback_manager=callback_manager,
   (...)
      8     streaming=True, 
      9 )
     10 llama = LangchainLLM(llama_lc)
---> 11 dd = SmartDataframe(df, config={"llm": llama_lc})
     13 dd.chat("scatter plot of sepallength vs sepalwidth color by class")

File ~/miniconda3/envs/data_agent/lib/python3.9/site-packages/pandasai/smart_dataframe/__init__.py:70, in SmartDataframe.__init__(self, df, name, description, config, logger)
     66 self._load_df(df)
     68 self._load_engine()
---> 70 self._dl = SmartDatalake([self], config=config, logger=logger)

File ~/miniconda3/envs/data_agent/lib/python3.9/site-packages/pandasai/smart_datalake/__init__.py:75, in SmartDatalake.__init__(self, dfs, config, logger, memory)
     66 """
     67 Args:
     68     dfs (List[Union[DataFrameType, Any]]): List of dataframes to be used
     69     config (Config, optional): Config to be used. Defaults to None.
     70     logger (Logger, optional): Logger to be used. Defaults to None.
     71 """
...
     32     if llm is None or not isinstance(llm, LLM or LangchainLLM):
---> 33         raise LLMNotFoundError("LLM is required")
     34     return llm

LLMNotFoundError: LLM is required

------------------------------------------------------------

Below works
```python
from pandasai import SmartDataframe
llama_lc = LlamaCpp(
    model_path="./wizardcoder-python-34b-v1.0.Q4_K_S.gguf",
    callback_manager=callback_manager,
    verbose=False,n_ctx=4096*4,
    n_gpu_layers=20,
    n_batch=20,
    streaming=True, 
)
llama = LangchainLLM(llama_lc)
dd = SmartDataframe(df, config={"llm": llama})

dd.chat("scatter plot of sepallength vs sepalwidth color by class")
```

**Comment by gventuri** [comment]
@hhy-joseph should be fixed with #512! Let me know if it works. Closing!

**Comment by emonzies** [comment]
@gventuri Yes it works for me now: switching from PandasAI class to SmartDataframe solve the pb with an AzureOpenAI langchain LLM
